File: protocol/scripts/testing/orders/test_short_term_grpc.js
#!/usr/bin/env node
/**
 * Tradeview Short-Term Order Submission via gRPC
 * Places orders directly into MemClob using the Tradeview v4 client
 */

const { CompositeClient, Network, OrderFlags, Order_Side, Order_TimeInForce } = require('@dydxprotocol/v4-client-js');
const { LocalWallet } = require('@dydxprotocol/v4-client-js');
const Long = require('long');

// Configuration - Tradeview explicit endpoints
const NETWORK = new Network(
    'testnet',
    'tradeview',
    'http://localhost:26657'
);
NETWORK.chainId = 'localtradeview';
NETWORK.validatorUrl = 'http://localhost:26657';
NETWORK.restUrl = 'http://localhost:1317';
NETWORK.indexerUrl = 'http://34.219.179.201:3002';
const ALICE_MNEMONIC = process.env.ALICE_MNEMONIC || "merge panther lobster crazy road hollow amused security before critic about cliff exhibit cause coyote talent happy where lion river tobacco option coconut small";
const BOB_MNEMONIC = process.env.BOB_MNEMONIC || "color black kingdom feed sure spread team month grow ill banner speed decade rule office burst uncle pure pump battle ready alone what taxi";

async function placeShortTermOrder() {
    console.log("=========================================");
    console.log("  Tradeview SHORT-TERM Order via gRPC");
    console.log("=========================================\n");

    try {
        // Initialize clients
        console.log("Step 1: Initializing clients...");
        // Updated for TradeView rebranding
        const aliceWallet = await LocalWallet.fromMnemonic(ALICE_MNEMONIC, 'tradeview');
        const bobWallet = await LocalWallet.fromMnemonic(BOB_MNEMONIC, 'tradeview');

        const aliceClient = await CompositeClient.connect(NETWORK);
        const bobClient = await CompositeClient.connect(NETWORK);

        console.log(`   Alice address: ${aliceWallet.address}`);
        console.log(`   Bob address: ${bobWallet.address}\n`);

        // Get current block height for GTB
        console.log("Step 2: Getting current block height...");
        const height = await aliceClient.validatorClient.get.latestBlockHeight();
        const goodTilBlock = height + 10; // 10 blocks from now

        console.log(`   Current height: ${height}`);
        console.log(`   Good Til Block: ${goodTilBlock}\n`);

        // Place Alice's BUY order
        console.log("Step 3: Placing Alice's SHORT-TERM BUY order...");
        const aliceOrder = {
            clientId: Math.floor(Date.now() / 1000), // Unix timestamp as client ID
            orderFlags: OrderFlags.SHORT_TERM, // SHORT-TERM order flag
            clobPairId: 0, // BTC-USD
            side: Order_Side.SIDE_BUY,
            quantums: Long.fromNumber(10_000_000), // 0.1 BTC
            subticks: Long.fromNumber(5_000_000), // Price: $50
            goodTilBlock: goodTilBlock,
            timeInForce: Order_TimeInForce.TIME_IN_FORCE_UNSPECIFIED,
            reduceOnly: false,
            clientMetadata: 0,
        };

        const aliceTx = await aliceClient.placeShortTermOrder(
            aliceWallet,
            aliceOrder,
            0 // subaccount number
        );

        console.log(`   ✅ Alice order placed!`);
        console.log(`   TxHash: ${aliceTx.hash}\n`);

        // Wait a bit
        await new Promise(resolve => setTimeout(resolve, 2000));

        // Place Bob's SELL order
        console.log("Step 4: Placing Bob's SHORT-TERM SELL order...");
        const bobOrder = {
            clientId: Math.floor(Date.now() / 1000) + 1,
            orderFlags: OrderFlags.SHORT_TERM,
            clobPairId: 0,
            side: Order_Side.SIDE_SELL,
            quantums: Long.fromNumber(10_000_000),
            subticks: Long.fromNumber(5_000_000), // Same price - will match!
            goodTilBlock: goodTilBlock,
            timeInForce: Order_TimeInForce.TIME_IN_FORCE_UNSPECIFIED,
            reduceOnly: false,
            clientMetadata: 0,
        };

        const bobTx = await bobClient.placeShortTermOrder(
            bobWallet,
            bobOrder,
            0
        );

        console.log(`   ✅ Bob order placed!`);
        console.log(`   TxHash: ${bobTx.hash}\n`);

        // Wait for inclusion
        console.log("Step 5: Waiting for order inclusion...");
        await new Promise(resolve => setTimeout(resolve, 5000));

        // Check if orders matched
        console.log("Step 6: Checking order status...");
        // Note: You'd query the indexer or check subaccount positions here

        console.log("\n=========================================");
        console.log("✅ SHORT-TERM ORDERS SUBMITTED VIA gRPC!");
        console.log("=========================================\n");
        console.log("Key Points:");
        console.log("  • Orders submitted directly to MemClob");
        console.log("  • Bypassed PrepareProposal filtering");
        console.log("  • Used SHORT_TERM order flags");
        console.log("  • Orders handled in-memory only");
        console.log("");

    } catch (error) {
        console.error("❌ Error:", error.message);
        console.error(error);
        process.exit(1);
    }
}

// Run
placeShortTermOrder().catch(console.error);

File: protocol/scripts/testing/orders/test_short_term_grpc.py
#!/usr/bin/env python3
"""
Short-Term Order Submission via gRPC
Alternative Python implementation using grpcurl
"""

import subprocess
import json
import time
from datetime import datetime

def get_current_height():
    """Get current block height"""
    result = subprocess.run(
        ['curl', '-s', 'http://localhost:26657/status'],
        capture_output=True,
        text=True
    )
    status = json.loads(result.stdout)
    return int(status['result']['sync_info']['latest_block_height'])

def place_order_grpc(address, subaccount, client_id, clob_pair, side, quantums, subticks, gtb):
    """
    Place short-term order using grpcurl to gRPC endpoint
    
    This bypasses the transaction mempool and goes directly to MemClob
    """
    
    order_json = {
        "order": {
            "order_id": {
                "subaccount_id": {
                    "owner": address,
                    "number": subaccount
                },
                "client_id": client_id,
                "order_flags": 0,  # SHORT_TERM flag
                "clob_pair_id": clob_pair
            },
            "side": side,  # 1=BUY, 2=SELL
            "quantums": str(quantums),
            "subticks": str(subticks),
            "good_til_block": gtb,
            "time_in_force": 0,
            "reduce_only": False,
            "client_metadata": 0
        }
    }
    
    # Use grpcurl to call the gRPC endpoint
    cmd = [
        'grpcurl',
        '-plaintext',
        '-d', json.dumps(order_json),
        'localhost:9090',  # gRPC port
        'dydxprotocol.clob.MsgService/PlaceOrder'
    ]
    
    print(f"   Calling gRPC endpoint...")
    result = subprocess.run(cmd, capture_output=True, text=True)
    
    if result.returncode == 0:
        print(f"   ✅ Order placed successfully!")
        print(f"   Response: {result.stdout}")
        return True
    else:
        print(f"   ❌ Error: {result.stderr}")
        return False

def main():
    print("=========================================")
    print("  Short-Term Orders via gRPC (Python)")
    print("=========================================\n")
    
    # Configuration
    ALICE = "tradeview199tqg4wdlnu4qjlxchpd7seg454937hj39sxzy"
    BOB = "tradeview10fx7sy6ywd5senxae9dwytf8jxek3t2g22s7jp"
    
    print("Step 1: Getting current block height...")
    current_height = get_current_height()
    gtb = current_height + 10
    print(f"   Current height: {current_height}")
    print(f"   Good Til Block: {gtb}\n")
    
    # Generate client IDs
    client_id_alice = int(time.time())
    client_id_bob = client_id_alice + 1
    
    print("Step 2: Placing Alice's BUY order via gRPC...")
    place_order_grpc(
        address=ALICE,
        subaccount=0,
        client_id=client_id_alice,
        clob_pair=0,
        side=1,  # BUY
        quantums=10_000_000,
        subticks=5_000_000,
        gtb=gtb
    )
    
    time.sleep(2)
    
    print("\nStep 3: Placing Bob's SELL order via gRPC...")
    place_order_grpc(
        address=BOB,
        subaccount=0,
        client_id=client_id_bob,
        clob_pair=0,
        side=2,  # SELL
        quantums=10_000_000,
        subticks=5_000_000,
        gtb=gtb
    )
    
    print("\n=========================================")
    print("✅ Orders submitted via gRPC!")
    print("=========================================\n")
    print("Note: This method:")
    print("  • Bypasses transaction mempool")
    print("  • Goes directly to MemClob")
    print("  • Avoids PrepareProposal filtering")
    print("  • Is the CORRECT way for short-term orders")
    print("")

if __name__ == "__main__":
    main()
